### PR TITLE
Saved search 2140

### PR DESCRIPTION
**Pull request by prakhyatox**
Ticket: https://github.com/FAIRsharing/fairsharing.github.io/issues/2140

### CODE REVIEW

**Comment by prakhyatox** [comment]
To do list: 
- [x] Saved search accessible from advancedSearch page
- [x] Saved search/es visible on the user's account page when logged in
- [x] Edit and delete action is completed on the user's account page
- [x] Saved search/es visible on the user's account page when logged out without actions column
- [x] Organisation and Record column in saved search table on the user's account page 
- [x] Show saved search on respective policy page if selected
- [x] Show saved search on respective organisation page if selected
- [x] Super power to super curators to delete any saved search of any user from any location (i.e. profile, policy , organisation)
- [x] Deletion of the saved search will delete its existence from FAIRsharing i.e. from user account's. organisation page, record's page
- [ ] Unit test cases

**Comment by allysonlister** [comment]
## Autocomplete
why does 'andersc' for a user search not return https://fairsharing.org/users/5015 ? other partial matches within a username seem to work (e.g. 'anderson' matches 'hataanderson' but WEIRDLY only when you type out anderson completely:
image

When you type out 'ander' you are not offered 'hataanderson' but instead this:
image


I realise this is a function that is not unique to the saved search, but the question remains though may be out of scope here.


I can search as follows to get this user:
image

But adding another character loses the user:
image

It just doesn't seem consistent.
> ** Prakhyat - Reply** : This is same functionality as on userslist page i.e. https://fairsharing.org/profiles/userslist. A new ticket has to be created to update this functionality in the site.
> ** Prakhyat - Reply 14/06/2024** : This has been fixed https://github.com/FAIRsharing/FAIRsharing-API/pull/998.

## 'link' link
When I click on the link it takes me to preview.fairsharing.org so I can't complete the test of creating a new saved search via clicking on the link as the feature isn't in preview.fairsharing.org - just to say this check is still outstanding. --- 

> **Prakhyat - Reply** : This is fixed now. Link will redirect to respective host environment


## Edit action
The edit action prints a warning about it creating a new search, then after that point the behaviour is the same as the "Link" link.
Perhaps we should modify this to say that the "Edit" action only allows you to edit the Name/Comment, with a note in the popup for doing the edit that to modify a saved search parameters, first open the existing saved search, then make the necessary updates to the search, then save as a different saved search....

When I say "ok" in the edit popup to open the search it takes me to preview.fairsharing.org so I can't complete the test of creating a new saved search via clicking on the link as the feature isn't in preview.fairsharing.org - just to say this check is still outstanding.

> **Prakhyat - Reply** : Link will redirect to respective host environment. But only editing Name/comment will add more complexity. Please confirm if only Name/Comments needs to be editable in edit action.

## save action
while I realised that you haven't yet implemented the visibility of the saved search on organisation or policy pages, I see also that is not yet implemented on other users' pages. I saved a search, and it correctly appeared in my user profile, but I had also asked it to be saved to lindsey's profile (user 5015 from above) and it is not there. Weirdly, when looking at https://deploy-preview-2299--fairsharing.netlify.app/users/5015 there isn't even a box visible that shows (empty) saved searches?
> **Prakhyat - Reply** : Assigned saved search is visible in the users' logged in profile. Viewing saved search in public profile has to be implemented

deleting a search works. 

I think that covers the current functionality, thanks!

**Comment by allysonlister** [comment]
One thing missing from the saved search box in the user account is the list of other places that saved search is present in. Therefore we would need another column with a list of records and organisations where the saved search has been linked. This will allow the user to go to those records/organisations to review that content. Note that only supercurators can have saved searches linked to organisations right now.

**Comment by allysonlister** [comment]
When deleting a saved search the popup should include an additional warning that this will remove the saved search from all of its linked locations, where applicable.

**Comment by prakhyatox** [comment]
> One thing missing from the saved search box in the user account is the list of other places that saved search is present in. Therefore we would need another column with a list of records and organisations where the saved search has been linked. This will allow the user to go to those records/organisations to review that content. Note that only supercurators can have saved searches linked to organisations right now.

This will be new addition to columns as it is not mentioned in gitbook documentation.

**Comment by prakhyatox** [comment]
> When deleting a saved search the popup should include an additional warning that this will remove the saved search from all of its linked locations, where applicable.

Added in Todo list https://github.com/FAIRsharing/fairsharing.github.io/pull/2299#issuecomment-2115537763

**Comment by allysonlister** [comment]
> Prakhyat - Reply : Assigned saved search is visible in the users' logged in profile. Viewing saved search in public profile has to be implemented

Does this mean "will be implemented" or does it mean "tell me if you would like it implemented?  
>_Prakhyat - Reply : This has now been implemented in public profile page_

> Prakhyat - Reply : Link will redirect to respective host environment. But only editing Name/comment will add more complexity. Please confirm if only Name/Comments needs to be editable in edit action.

What I'm saying is that the 'Edit' and  'link' are, right now, essentially performing the same function as both simply redirect you to the search where (if desired) you can adjust and save as a new search. This is not the expected behaviour. I suggest the following:

1. Make both the name and comment clickable - when clicked, they will bring up a popup or just make that field editable by the user.
2. change the 'edit' icon behaviour to instead open up the advanced search window pre-filled with the values within that saved search. the user will naturally save as a separate search if they choose to do so after modifying the values.
>_Prakhyat - Reply : Action item has been updated, with easier and similar UI as per the other action item on the profile page._

> ** Prakhyat - Reply** : This is same functionality as on userslist page i.e. https://fairsharing.org/profiles/userslist. A new ticket has to be created to update this functionality in the site.

Please could you do so?
>_Prakhyat - Reply : Ticket is created https://github.com/FAIRsharing/FAIRsharing-API/issues/997._

**Comment by allysonlister** [comment]
Screenshot from 2024-06-12 11-18-50
Thinking more about what the word "Select" means in the step names (e.g. "Select Policy"), I think it might be confusing. Apologies, but please could we replace Select with...

Create Policy Link; Enter text to search for policy record(s) to associate with this saved search
Create Organisation Link; Enter text to search for organisation(s) to associate with this saved search
Create User Link; Enter text to search for additional user(s) to associate with this saved search

**Comment by prakhyatox** [comment]
> Screenshot from 2024-06-12 11-18-50 Thinking more about what the word "Select" means in the step names (e.g. "Select Policy"), I think it might be confusing. Apologies, but please could we replace Select with...
> 
> Create Policy Link; Enter text to search for policy record(s) to associate with this saved search Create Organisation Link; Enter text to search for organisation(s) to associate with this saved search Create User Link; Enter text to search for additional user(s) to associate with this saved search

Done

File: src/components/Records/Record/GeneralInfo/SavedSearches.vue [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+    <!--        text-color="white"-->
+    <!--        @click:close="confirmDeleteSavedSearch(search)"-->
+    <!--      >-->
+    <!--        <a-->
+    <!--          class="white&#45;&#45;text"-->
+    <!--          :href="search.url"-->
+    <!--        >-->
+    <!--          {{ search.name }}-->
+    <!--        </a>-->
+    <!--      </v-chip>-->
```
Comment (knirirr) [inline src/components/Records/Record/GeneralInfo/SavedSearches.vue]: Is this block something which might be needed in future?

File: src/components/Records/Search/SaveSearch/StepperComponents/ResultCard.vue [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+  >
+    <v-card-title
+      class="justify-center white--text mb-4"
+      style="background-color: green"
+    >
+      Success !!
+    </v-card-title>
+    <v-card-text class="pb-0">
+      <p class="text-body-1">
+        Your search is saved successfully. Please check saved search table under
```
Comment (knirirr) [inline src/components/Records/Search/SaveSearch/StepperComponents/ResultCard.vue]:
```suggestion
        Your search was saved successfully. Please check the saved search table under
```

**Comment by allysonlister** [comment]
This is looking AMAZING! I have a few items, but the search overall is working really really well. Thank you!

- [x] The users I have added the saved search to from my supercurator account do not appear in my user profile in the same way that the organisations and policies do. In the below example I cannot remember about the older saves, but certainly the 24.6 saved search includes lnanderscience as an additional user. 
image
- [x] As we have decided the functionality of 'edit search' is to take the user to the search result page so they can save it with a different name and comment should they so desire, we need to update things. Here's what we should do:
1. Remove the 'Edit Search' option completely
2. use a bin icon instead of 'Delete Search' and place it where the ":" currently is at the end of the row (I know there are three dots shown, but I don't know what that's called).
3. Add an explanatory sentence just below the blue "Saved Searches" row with this text: "Clicking on the name of a saved search will take you to its search results. From the results page, if you are logged in you may further refine the search and/or save the search yourself. More information on Conforming Resources and Saved Searches can be found in our [Gitbook documentation](https://fairsharing.gitbook.io/fairsharing/how-to/advanced-search)."
- [x] As shown in the screenshot below, even when I am logged in as the unprivileged allyson_user account, I may still delete Lindsey's saved searches! Only Lindsey or any supercurator should have access to deletion of searches in her profile page.
image 

> Prakhyat-reply- As per the screenshot you are logged in as allysonlister ( please see creator column). It does not shows allyson_user in the creator column.

- [x] Add a hyperlinked creator username to the white saved search inset box in Conforming Resources:
image
- [x] On policy pages, please could you slightly update the text describing the Conforming Resources tab to this: "This policy has certain requirements and recommendations that have been used to create a conformant FAIRsharing search. The searches listed within this section will provide an initial filtering of the FAIRsharing registries. Clicking on a saved search will take you to a set of search results which you may further refine according to your needs and any additional policy requirements. If you are logged in, you may also save any search results to your own profile. More information on Conforming Resources and Saved Searches can be found in our Gitbook documentation. "
- [ ] Deletion behaviour:
1. deleting a search from a policy should ONLY delete the search from that policy; at the moment it is deleting it from everywhere.
image
2. deleting a search from an organisation should ONLY delete the search from that organisation; at the moment it is deleting it from everywhere.
3. CORRECT: deleting a search from the OWNING user's profile should delete it everywhere.
4. when deleting an item from the OWNING user's profile, the popup should match the others (see example just above) and include the warning about it being deleted from everywhere. At the moment it doesn't have that warning:
image
5. When deleting a search from a LINKED user's profile (not the owner), the search should ONLY be deleted from that linked user's profile; at the moment it is deleting it from everywhere.

**Comment by knirirr** [comment]
> As shown in the screenshot below, even when I am logged in as the unprivileged allyson_user account, I may still delete Lindsey's saved searches! Only Lindsey or any supercurator should have access to deletion of searches in her profile page.

This should not happen if the server is correctly applying its SavedSearchPolicy, and that's where the invesitgation will have to take place.

File: src/components/Records/Search/SaveSearch/StepperComponents/ResultCard.vue (hunk 2 above)
Comment (prakhyatox) [reply src/components/Records/Search/SaveSearch/StepperComponents/ResultCard.vue]: Fixed

File: src/components/Records/Record/GeneralInfo/SavedSearches.vue (hunk 1 above)
Comment (prakhyatox) [reply src/components/Records/Record/GeneralInfo/SavedSearches.vue]: Not needed anymore, hence removed.

**Comment by allysonlister** [comment]
Note that there is one other place in the specification that is different from the implementation with regards to deletion:

> Deletion of a saved search in a user profile, organisation or policy record should happen via that object's edit interface. At the moment, only curators can edit an organisation so they will be the only ones who can delete the saved searches (this is a little odd as any member of an organisation can add a saved search, but I think it will be ok for now).

Right now, deletion does NOT happen via that object's edit interface but rather via a bin icon that appears only to those with the correct permissions.

While I am happy - in theory - to keep this, there is one issue with this (and is why I didn't ask for it this way in the first place): the addition and removal of the linked searches will NOT trigger a change in the curation dashboard for policies, nor will it be visible in the record history. Both of these things are functions I wanted to have. (I realise I wasn't explicit about this wrt **addition** of searches, so I'm not necessarily asking to enable it now)

Ideally, addition of the linked searches to organisations, user accounts and policies should happen exactly as they are happening in this PR. However, for _policies_ they should trigger an update in the edit history and a row in the curation dashboard (this is why I wrote the specification in the way I did)

Then, in removal of linked searches for _policies_ only, such removals should trigger an update of the record that is recorded in the edit history and a row in the curation dashboard (this is why I wrote the specification in the way I did).

If this is too much, keep it all the way it is. But I just wanted to point this out as a part of the spec that different from how it looks. I also realise I wasn't completely clear.

**Comment by allysonlister** [comment]
I haven't been able to check the user profile page as for some reason it's hanging when I'm trying to load https://deploy-preview-2299--fairsharing.netlify.app/accounts/profile 

But I did save a search and found some odd behaviour around the addition of more than one user and more than one org; I don't think I tried more than one user the other day, so this may be a bug that's been around for a while. Adding multiple policies seems ok, but orgs display very oddly (as though they're disappearing) though they seem to be added ok; with users it somehow deletes the previous one I've added when I search for the second:
[Screencast from 2024-06-25 17-53-32.webm](https://github.com/FAIRsharing/fairsharing.github.io/assets/143586/4cda7f1e-dd58-4032-b629-e25b066673d6)

**Comment by allysonlister** [comment]
Looking great! 

- [x] major: unlinking a saved search from an 'other' user **deletes** the saved search from everywhere incl the owning user's account (rather than just unlinking from that 'other' user's account). In such cases, unlinking from an 'other' user should ONLY unlink it from that user. It is ONLY when you delete from the OWNING user's account that it deletes from everywhere. It may be that you then use the 'bin' icon only on the owning user's saved searches, and if a search is saved to an 'other' user perhaps the icon should be the 'unlink' icon?
- [x] major: when I unlink a saved search from an organisation, all other organisations that have that saved search are ALSO unlinked. It should be only removed from the specified organisation.
- [x] minor: Update the Saved Search explanatory text that is within the user profile table to "Clicking on the name of a saved search will take you to its search results. From the results page, if you are logged in you may further refine the search and/or save the search yourself. More information on ~~Conforming Resources and ~~Saved Searches can be found in our [Gitbook documentation](https://fairsharing.gitbook.io/fairsharing/how-to/advanced-search)."
- [x] minor: Please could you put a newline in here? Or at least a space? --- _Could not replicate this behvaiour in any screen dimension._

image

_Reply_:
image



- [x] minor: When I look at Lindsey's profile, where I added a policy saved search, the explanatory sentence is NOT present in this 'other' user's table below the words "saved search" (https://deploy-preview-2299--fairsharing.netlify.app/users/5015)
image

**Comment by allysonlister** [comment]
- [x] Since I last commented (and this may be something to do with Firefox or my OS), something weird is happening when I select multiple user accounts. If you can see from this screencast, it 'removes' previously-selected accounts when I start typing a new account name. The correct set of users are saved, so it is a display rather than functionality issue.

Firefox 129.0.2 (64-bit) - snap for Ubuntu
[Screencast from 2024-08-27 13-23-47.webm](https://github.com/user-attachments/assets/9a5505b7-822e-45f2-a337-dabb35c81180)


Chromium (Version 128.0.6613.84 (Official Build) snap (64-bit)) appears to have the same behaviour:
[Screencast from 2024-08-27 13-29-57.webm](https://github.com/user-attachments/assets/262118ff-0410-4536-805f-5c46b2b9b5ff)

I'm pretty sure previous iterations of the saved search did correctly display multiple users, but I'm not 100% sure.

- [x] BTW I didn't mean for you to add "**~~**" to the actual explanatory text; I was just trying to highlight the modified text for you. Please could you remove the "**~~**"?


- [x] Finally, wrt the newline in the user profile explanatory text: it seems to be a firefox issue. The two screenshots below are with Chromium and Firefox; you can see the newline issue only happens in the firefox screenshot.

Screenshot from 2024-08-27 13-40-06
Screenshot from 2024-08-27 13-40-11


Can you check and see if you can find a fix for firefox please?

**Review by allysonlister — APPROVED** [review]
Looks great, thank you - all of my known issues are resolved. Amazing work!

**Pull request merged**